### DISCUSSION TITLE
github actions repository dispatch "message": "Problems parsing JSON"

### DISCUSSION THREAD

**Discussion by riphunter3141** [community/community#28224] [Actions] [answered]
I am trying to create an actions workflow that can be triggered from an external event. Upon researching, found out I can do it through repository_dispatch.

I need to trigger the actions through python but I get this error and status code 400:

> {'message': 'Problems parsing JSON', 'documentation_url': 'https://docs.github.com/rest/reference/repos#create-a-repository-dispatch-event'}

I then serialized the payload dictionary to string with json.dumps() and the error changed to with code 422:

> {'message': 'Invalid request.\n\nFor \'links/0/schema\', "{\\"event_type\\": \\"test\\", \\"client_payload\\": {\\"unit\\": false, \\"integration\\": true}}" is not an object.', 'documentation_url': 'https://docs.github.com/rest/reference/repos#create-a-repository-dispatch-event'}

Here's my code snippet that I constructer from the curl example provided in the docs
```
import requests
import json

url = "https://api.github.com/repos/larwindcunha/<repo>/dispatches"
payload = {"event_type": "test", "client_payload": {"unit":False,"integration":True}}
header = {"Accept": "application/vnd.github+json", "Authorization": "token <my_token>"}
payload = json.dumps(payload)
resp = requests.post(url=url, headers=header, json=payload)
```

When I try running the curl command provided here - https://docs.github.com/en/rest/repos/repos#create-a-repository-dispatch-event, I get the same error with code 400:

> {'message': 'Problems parsing JSON', 'documentation_url': 'https://docs.github.com/rest/reference/repos#create-a-repository-dispatch-event'}

```
curl \
  -X POST \
  -H "Accept: application/vnd.github+json" \ 
  -H "Authorization: token <TOKEN>" \
  https://api.github.com/repos/OWNER/REPO/dispatches \
  -d '{"event_type":"on-demand-test","client_payload":{"unit":false,"integration":true}}'
```

Do you guys know if I'm doing something wrong here? Or is repository dispatch broken? Any help would be appreciated.

**Comment by airtower-luna** [comment, marked as answer]
Here's your problem:
```py
payload = json.dumps(payload)
resp = requests.post(url=url, headers=header, json=payload)
```
`json.dumps()` expects a Python data structure, serializes that as JSON, and returns it as a string. The `json` parameter for `requests.post()` expects a Python data structure, serializes that as JSON, and sends it as the post body. Seems familiar? The result is that you get JSON for a string that contains JSON.

Remove the `payload = json.dumps(payload)` line and it should work. :smile_cat:

**Reply by riphunter3141** [reply]
Thanks, your solution is accurate and as it turns out, I did try this out before posting my issue here but received an error which misled me.

> Traceback (most recent call last):
>   File "C:\Users\dcunh\Documents\Code\Test\env_test\lib\site-packages\requests\models.py", line 971, in json
>     return complexjson.loads(self.text, **kwargs)
>   File "C:\Users\dcunh\AppData\Local\Programs\Python\Python39\lib\json\__init__.py", line 346, in loads
>     return _default_decoder.decode(s)
>   File "C:\Users\dcunh\AppData\Local\Programs\Python\Python39\lib\json\decoder.py", line 337, in decode
>     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
>   File "C:\Users\dcunh\AppData\Local\Programs\Python\Python39\lib\json\decoder.py", line 355, in raw_decode
>     raise JSONDecodeError("Expecting value", s, err.value) from None
> json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "C:\Users\dcunh\Documents\Code\Test\test.py", line 10, in <module>
>     print(resp.json())
>   File "C:\Users\dcunh\Documents\Code\Test\env_test\lib\site-packages\requests\models.py", line 975, in json
>     raise RequestsJSONDecodeError(e.msg, e.doc, e.pos)
> requests.exceptions.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

The reason I missed out on the solution was that I was trying to print(resp.json()) since I was getting a json response on error case, but as it turns out, on successful post request there was no response. I just got a JSON decoder error. This misled me as I assumed that some json data would be sent back in the response for positive case as it was for a negative one. Later when I checked my actions log, it was being triggered.
If github decides to send data for unsuccessful cases, then it should send data for successful case as well. Or atleast put out a disclaimer so this can be taken into account in error handling.

Thanks again!

**Reply by airtower-luna** [reply]
Technically they mention it by sending a 204 No Content response, but that might not be super obvious. I'd probably just use `resp.raise_for_status()` to see if there was a problem. :grin:

**Comment by iHarshad** [comment]
I made the changes that @airtower-luna suggested, but the error persists. In the remote repo workflow file I have enabled both `workflow_dispatch` and `repository_dispatch`. I want to go with `workflow_dispatch` which is more [flexible](https://stackoverflow.com/a/68148239).

<details>
  <summary>Python + workflow_dispatch</summary>
  
  

  ```python
    
  import requests
  
  GH_USER = "<user>"
  GH_REPO = "<repo>"
  GH_PAT = "<my token>"
  GH_WORKFLOW = "<workflow.yml>"
  
  BASE_URL = "https://api.github.com"
  API_URL = (
    f"{BASE_URL}/repos/{GH_USER}/{GH_REPO}/actions/workflows/{GH_WORKFLOW}/dispatches"
  )
  
  request_data = {
    "ref": "master"
  }
  
  request_headers = {
    "Accept": "application/vnd.github+json",
    "Authorization": f"Bearer {GH_PAT}",
    "X-GitHub-Api-Version": "2022-11-28",
  }
  
  try:
    response = requests.post(
      API_URL, headers=request_headers, json=request_data, allow_redirects=True
    )
  
    # Check if the request was successful (status code 2xx)
    response.raise_for_status()
  
    print(response.json())
  
  except requests.exceptions.RequestException as e:
    print(f"Error during API call: \n{e}")
  ```
  
  

**Output:**
  ```
  Error during API call: 
  Expecting value: line 1 column 1 (char 0)
  ```

</details>



<details>
  <summary>When I use `curl` + workflow_dispatch it works properly:</summary>
  
  
  ```shell
  curl -X POST --write-out 'Response Status: %{http_code}' \
    -fsSL "https://api.github.com/repos/<GH_USER>/<GH_REPO>/actions/workflows/<WORKFLOW.yml>/dispatches" \
    -H "Accept: application/vnd.github+json" \
    -H "Authorization: Bearer <TOKEN>" \
    -H "X-GitHub-Api-Version: 2022-11-28" \
    --data '{"ref": "master"}'  
```
  
  
**Output:**
  ```
  Response Status: 204
  ```

</details>



<details>
  <summary>I also tried the original code with minor changes and got the same error:</summary>
  
  

  ```python
  import requests
  import json
  
  GH_USER = "<user>"
  GH_REPO = "<repo>"
  GH_PAT = "<my token>"
  GH_WORKFLOW = "<workflow.yml>"
  
  url = f"https://api.github.com/repos/{GH_USER}/{GH_REPO}/dispatches"
  payload = {
      "event_type": "test",
      "client_payload": {
          "unit": False,
          "integration": True
      }
  }
  header = {
      "Accept": "application/vnd.github+json", 
      "Authorization": f"token {GH_PAT}"
  }
  
  try:
      resp = requests.post(url=url, headers=header, json=payload)
  
      resp.raise_for_status()
  
      print(resp.json())
  
  except requests.exceptions.RequestException as e:
      print(f"Error during API call: \n{e}")
  ```  

</details>

**Reply by airtower-luna** [reply]
Remove the try/except or use [`traceback`](https://docs.python.org/3/library/traceback.html) in the `except` block to see the full stack trace. With just that output it's hard to say what might be wrong, to put it mildly. :sweat_smile:

**Reply by iHarshad** [reply]
Thanks for the help @airtower-luna 

The problem was in the print statement for the API response. As with a usual API JSON response, you try to print the response body. Here the Github API call only returns a `204` status code and no response body.

<details>
  <summary>Updated code</summary>
  
  **TODO:** Update `GH_USER`, `GH_REPO`, `GH_PAT` and`GH_WORKFLOW` variables. 
  
  ```python
    import requests
    import traceback

    GH_USER = "<user>"
    GH_REPO = "<repo>"
    GH_PAT = "<my token>"    # Github Personal Access Token
    GH_WORKFLOW = "<workflow.yml>"
    
    BASE_URL = "https://api.github.com"
    API_URL = (
        f"{BASE_URL}/repos/{GH_USER}/{GH_REPO}/actions/workflows/{GH_WORKFLOW}/dispatches"
    )
    
    request_data = {
        "ref": "master"
    }

    request_headers = {
        "Accept": "application/vnd.github+json",
        "Authorization": f"Bearer {GH_PAT}",
        "X-GitHub-Api-Version": "2022-11-28",
    }

    try:
        response = requests.post(
            API_URL, headers=request_headers, json=request_data, allow_redirects=True
        )

        # Check if the request was successful (status code 2xx)
        response.raise_for_status()

        if response.status_code == 204 and response.ok:
            print(f"SUCCESS\nResponse = {response}")
    
    except requests.exceptions.RequestException as e:
        print(f"Error during API call: \n{e}\n\n")
        traceback.print_exc()
  ```
  
</details>

<details>
  <summary>Success Output</summary>
  
  ```
  SUCCESS
  Response = <Response [204]>
  ```
  
</details>

Now moving on to passing parameter values to `workflow_dispatch`. Most references on the internet use `repository_dispatch` for this which has a defined `client_payload` parameter. I think I could leverage the `inputs` parameter for this 🤞🏼.

**Reply by iHarshad** [reply]
Here is the updated workflow for future reference.

<details>
  <summary>Github Workflow File</summary>
  
  ```yaml
  ---
  # test.yml
  
  
  name: test
  
  on:
    workflow_dispatch:
      inputs:
        message1:
          description: 'Parameter message'     
          required: true
          default: 'no message'
        tag1:
          description: 'Test scenario tags'
          required: false
  
  jobs:
    printInputs:
      runs-on: ubuntu-latest
      steps:
        - run: |
            echo "Message: ${{ github.event.inputs.message1 }}"
            echo "Tags: ${{ github.event.inputs.tag1 }}"
  ```
 
</details>


<details>
  <summary>Python API Call Script</summary>
  
  ```python
  # run.py
  
  import requests
  import traceback

  GH_USER = "<user>"
  GH_REPO = "<repo>"
  GH_PAT = "<my token>"    # Github Personal Access Token
  GH_WORKFLOW = "test.yml"
  
  BASE_URL = "https://api.github.com"
  API_URL = (
      f"{BASE_URL}/repos/{GH_USER}/{GH_REPO}/actions/workflows/{GH_WORKFLOW}/dispatches"
  )
  
  request_data = {
      "ref": "master",
      "inputs": {
          "tag1": "api call",
          "message1": "triggered from python script"
      }
  }

  request_headers = {
      "Accept": "application/vnd.github+json",
      "Authorization": f"Bearer {GH_PAT}",
      "X-GitHub-Api-Version": "2022-11-28",
  }
  
  try:
      response = requests.post(
          API_URL, headers=request_headers, json=request_data, allow_redirects=True
      )
      # Check if the request was successful (status code 2xx)
      response.raise_for_status()
  
      if response.status_code == 204 and response.ok:
          print(f"SUCCESS\nresponse = {response}")
  
  except requests.exceptions.RequestException as e:
      print(f"Error during API call:")
      traceback.print_exc()  # Use traceback to print the full stack trace
  ```
  
</details>

If you want to use the API payload in a script (python/JS) then pass it as a param to the script. `MY_VAR=$GITHUB_XXX python3 my-script.py`. There is a ton of info about this in [GuillaumeFalourd/poc-github-actions](https://github.com/GuillaumeFalourd/poc-github-actions/blob/198c787f038466418e51c791b5b2ecde40cd03c3/.github/workflows/11-input-workflow.yml#L24)
